sympy/matrices/decompositions.py

```diff
@@ -0,0 +1,192 @@
+
+
+def _cholesky(M, hermitian=True, dotprodsimp=None):
+    """Returns the Cholesky-type decomposition L of a matrix A
+    such that L * L.H == A if hermitian flag is True,
+    or L * L.T == A if hermitian is False.
+
+    A must be a Hermitian positive-definite matrix if hermitian is True,
+    or a symmetric matrix if it is False.
+
+    Parameters
+    ==========
+
+    dotprodsimp : bool, optional
+        Specifies whether intermediate term algebraic simplification is used
+        during matrix multiplications to control expression blowup and thus
+        speed up calculation.
+
+    Examples
+    ========
+
+    >>> from sympy.matrices import Matrix
+    >>> A = Matrix(((25, 15, -5), (15, 18, 0), (-5, 0, 11)))
+    >>> A.cholesky()
+    Matrix([
+    [ 5, 0, 0],
+    [ 3, 3, 0],
+    [-1, 1, 3]])
+    >>> A.cholesky() * A.cholesky().T
+    Matrix([
+    [25, 15, -5],
+    [15, 18,  0],
+    [-5,  0, 11]])
+
+    The matrix can have complex entries:
+
+    >>> from sympy import I
+    >>> A = Matrix(((9, 3*I), (-3*I, 5)))
+    >>> A.cholesky()
+    Matrix([
+    [ 3, 0],
+    [-I, 2]])
+    >>> A.cholesky() * A.cholesky().H
+    Matrix([
+    [   9, 3*I],
+    [-3*I,   5]])
+
+    Non-hermitian Cholesky-type decomposition may be useful when the
+    matrix is not positive-definite.
+
+    >>> A = Matrix([[1, 2], [2, 1]])
+    >>> L = A.cholesky(hermitian=False)
+    >>> L
+    Matrix([
+    [1,         0],
+    [2, sqrt(3)*I]])
+    >>> L*L.T == A
+    True
+
+    See Also
+    ========
+
+    sympy.matrices.matrices.MatrixBase.LDLdecomposition
+    LUdecomposition
+    QRdecomposition
+    """
+
+    from .dense import MutableDenseMatrix
+
+    if not M.is_square:
+        raise NonSquareMatrixError("Matrix must be square.")
+    if hermitian and not M.is_hermitian:
+        raise ValueError("Matrix must be Hermitian.")
+    if not hermitian and not M.is_symmetric():
+        raise ValueError("Matrix must be symmetric.")
+
+    dps = _dotprodsimp if dotprodsimp else expand_mul
+    L   = MutableDenseMatrix.zeros(M.rows, M.rows)
+
+    if hermitian:
+        for i in range(M.rows):
+            for j in range(i):
+                L[i, j] = dps((1 / L[j, j])*(M[i, j] -
+                    sum(L[i, k]*L[j, k].conjugate() for k in range(j))))
+
+            Lii2 = dps(M[i, i] -
+                sum(L[i, k]*L[i, k].conjugate() for k in range(i)))
+
+            if Lii2.is_positive is False:
+                raise NonPositiveDefiniteMatrixError(
+                    "Matrix must be positive-definite")
+
+            L[i, i] = sqrt(Lii2)
+
+    else:
+        for i in range(M.rows):
+            for j in range(i):
+                L[i, j] = dps((1 / L[j, j])*(M[i, j] -
+                    sum(L[i, k]*L[j, k] for k in range(j))))
+
+            L[i, i] = sqrt(dps(M[i, i] -
+                sum(L[i, k]**2 for k in range(i))))
+
+    return M._new(L)
+
+
+def _cholesky_sparse(M, hermitian=True, dotprodsimp=None):
```

Comment:
I think that `hermitian=True` should be the decomposition `L * L.H` unlike documented here. 
And I don't think that it's a correct patch unless you add an algorithm to compute hermitian cholesky for sparse.

Reply:
Ok, for some stupid reason I had the impression the sparse algorithm could handle it but obviously not. Added the hermitian changes to the algorithm and now it behaves like dense:
```julia
>>> Matrix([[5 + I, 0], [0, 1]]).cholesky(hermitian=False)
Matrix([
[sqrt(5 + I), 0],
[          0, 1]])
>>> _ * _.T
Matrix([
[5 + I, 0],
[    0, 1]])

>>> SparseMatrix([[5 + I, 0], [0, 1]]).cholesky(hermitian=False)
Matrix([
[sqrt(5 + I), 0],
[          0, 1]])
>>> _ * _.T
Matrix([
[5 + I, 0],
[    0, 1]])

>>> Matrix([[2, 1 + I], [1 - I, 2]]).cholesky(hermitian=True)
Matrix([
[                sqrt(2), 0],
[sqrt(2)/2 - sqrt(2)*I/2, 1]])
>>> simplify (_ * _.H)
Matrix([
[    2, 1 + I],
[1 - I,     2]])

>>> SparseMatrix([[2, 1 + I], [1 - I, 2]]).cholesky(hermitian=True)
Matrix([
[                sqrt(2), 0],
[sqrt(2)/2 - sqrt(2)*I/2, 1]])
>>> simplify (_ * _.H)
Matrix([
[    2, 1 + I],
[1 - I,     2]])
```